cxx/relation.hh

```diff
@@ -23,21 +30,22 @@ class T_relation {
   // is a distribution over.
   std::vector<std::string> domains;
 
-  // Must be the name of a distribution in distributions/.
-  std::string distribution;
+  DistributionSpec distribution_spec;
 };
 
 template <typename DistributionType>
 class Relation {
  public:
   using ValueType = typename DistributionType::SampleType;
   using DType = DistributionType;
-  static_assert(std::is_base_of<Distribution<ValueType>, DType>::value, 
+  static_assert(std::is_base_of<Distribution<ValueType>, DType>::value,
                 "DistributionType must inherit from Distribution.");
   // human-readable name
   const std::string name;
-  // Distribution over the relation's codomain.
-  const std::string distribution;
+  // Relation spec.
+  T_relation trel;
+  // Distribution prior over the relation's codomain.
+  const DistributionType* cluster_prior;
```

Comment:
Rather than storing this as a DistributionType*, you could store the DistributionSpec and just use cluster_prior_from_spec to generate the new distributions when needed.  That way you wouldn't need to add the prior() method to all the distributions.

Reply:
Thanks for catching that! In an earlier iteration I was storing the whole distribution_args string in DistributionSpec, and I wanted to avoid re-parsing the string every time a new prior instance was needed, but now that DistributionSpec contains an argname/argval map I agree this is better.